docs/proposals/20210210-insulate-users-from-kubeadm-API-changes.md

```diff
@@ -0,0 +1,226 @@
+
+- Hosted in `bootstrap/kubeadm/types/v1beta2`.
+- Currently, not used in the Cluster API codebase.
+- Does not include changes for better CRD support introduced in kubeadm v1beta1
+  mirror-types.
+
+#### Cluster API v1alpha3 changes
+
+Changes to cluster API v1alpha3 release should be minimal and no breaking change
+should be introduced while implemeting this proposal.
+
+__Alternative 1:__
+
+Keep kubeadm v1beta1 types as a Hub type; implement conversion to kubeadm API
+version f(Kubernetes Version) when generating the kubeadm config for init/join.
+
+This alternative is the more clean, robust and forward looking, but it requires
+much more work than alternative 2.
+
+__Alternative 2:__
+
+Keep kubeadm v1beta1 types as a Hub type; only change the apiVersion to
+`kubeadm.k8s.io/v1beta2` in the generated kubeadm config for init/join.
+
+This alternative is based on following considerations:
+
+- kubeadm v1beta1 mirror types are "compatible" with kubeadm v1beta2 types
+  - Having support for IgnorePreflightErrors into the NodeRegistrationOptions
+    is not required for Cluster API v1alpha3
+  - The automatic certificate copy feature is not used by Cluster API
+  - Improved serialization support has been already applied to the kubeadm
+    v1beta1 mirror-types (and further extended).
+- the minimal Kubernetes version supported by Cluster API is v1.16, and kubeadm
+  v1.16 could work with v1beta2 types. Same for all the other versions up to latest(1.20)
+  and next (1.21).
+- limitations: this approach is not future proof, and it should be reconsidered
+  whenever a new version of kubeadm types is created while v1alpha3 is still supported.
+
+__Common considerations for both alternatives__
+
+KCP is modifying the kubeadm-config Config Map generated by kubeadm, and ideally also
+this bit of code should be made kubeadm version aware.
+
+However, given that the current implementation currently uses unstructured, and
+a change for using the kubeadm types, requires a big refactor, the proposed approach
+for v1alpha3 is to limit the changes to only upgrade the apiVersion when required.
+
+- limitations: this approach is not future proof, and it should be reconsidered
+  whenever a new version of kubeadm types is changing one of the fields
+  edited during upgrades.
+
+#### Cluster API v1alpha4 changes
+
+Changes to cluster API v1alpha4 could be more invasive and seek for a
+forward-looking solution.
+
+Planned actions are:
+
+- introduce a Cluster API owned version of the kubeadm config types
+  (starting from kubeadm v1beta1?) to be used KubeadmConfig/KubeadmControlPlane
+  specs; this should also act as a serialization/deserialization hub.
+- preserve `bootstrap/kubeadm/types/v1beta1` as a serialization/deserialization
+  spoke for v1alpha4 (also, this will be used by v1alpha3 API types until removal)
+- preserve `bootstrap/kubeadm/types/v1beta2` as serialization/deserialization
+  poke for v1alpha4
+- implement hub/spoke conversions
```

Comment:
what is a "spoke" version?

Reply:
this is referring to how controller-runtime handles conversion, see https://book.kubebuilder.io/multiversion-tutorial/conversion-concepts.html
(added link to the document as well)